async-nats/src/service/mod.rs

```diff
@@ -0,0 +1,278 @@
+        config.endpoint.clone().subject
+    );
+    let mut pings =
+        verb_subscription(client.clone(), Verb::Ping, config.name.clone(), id.clone()).await?;
+    let mut infos =
+        verb_subscription(client.clone(), Verb::Info, config.name.clone(), id.clone()).await?;
+    let mut schemas = verb_subscription(
+        client.clone(),
+        Verb::Schema,
+        config.name.clone(),
+        id.clone(),
+    )
+    .await?;
+    let mut status = verb_subscription(
+        client.clone(),
+        Verb::Status,
+        config.name.clone(),
+        id.clone(),
+    )
+    .await?;
+    let handle = tokio::task::spawn({
+        let client = client.clone();
+        let info_json = serde_json::to_vec(&info).map(Bytes::from)?;
+        let schema_json = serde_json::to_vec(&json!({
+            "name": config.name.clone(),
+            "id": id.clone(),
+            "version": config.version.clone(),
+            "schema": config.schema,
+        }))
+        .map(Bytes::from)?;
+        async move {
+            loop {
+                tokio::select! {
+                    Some(ping) = pings.next() => {
+                        println!("PING");
+                        client.publish(ping.reply.unwrap(), info_json.clone()).await?;
+                    },
+                    Some(info) = infos.next() => {
+                        println!("INFO");
+                        client.publish(info.reply.unwrap(), info_json.clone()).await?;
+                    },
+                    Some(schema) = schemas.next() => {
+                        println!("SCHEMA");
+                        client.publish(schema.reply.unwrap(), schema_json.clone()).await?;
+                    },
+                    // FIXME: proper status handling
+                    Some(status) = status.next() => {
+                        let stats = serde_json::to_vec(&json!({
+                            "name": "name",
+                            "stats": "stats",
+                        }))?;
+                        client.publish(status.reply.unwrap(), stats.into()).await?;
+                    },
+                    else => break,
+                }
+            }
+            Ok(())
+        }
+    });
+    Ok(Service {
+        stats,
+        id,
+        config,
+        info,
+        client,
+        requests,
+        handle,
+    })
+}
+
+async fn verb_subscription(
+    client: Client,
+    verb: Verb,
+    name: String,
+    id: String,
+) -> Result<futures::stream::Fuse<SelectAll<Subscriber>>, Error> {
+    let ping_all = client
+        .subscribe(format!("{}.{}", SERVICE_API_PREFIX, verb))
+        .await?;
+    let ping_name = client
+        .subscribe(format!("{}.{}.{}", SERVICE_API_PREFIX, verb, name))
+        .await?;
+    let ping_id = client
+        .subscribe(format!("{}.{}.{}.{}", SERVICE_API_PREFIX, verb, name, id,))
+        .await?;
+    Ok(stream::select_all([ping_all, ping_id, ping_name]).fuse())
+}
+
+impl Stream for Service {
```

Comment:
Rather than service being a stream, maybe it has a stream? that way for_each can track statistics and you can downcast into the raw stream (something ala into_stream)?